internal/cephfs/groupcontrollerserver.go

```diff
@@ -61,3 +72,566 @@ func (cs *ControllerServer) validateCreateVolumeGroupSnapshotRequest(
+			"source volume ids %v do not match in the existing volume group snapshot %v",
+			req.GetSourceVolumeIds(),
+			vgs.GetVolumeIDs())
+	}
+	// Release the quiesce of the subvolumes and subvolume groups in the
+	// filesystems for the volumes.
+	for _, fm := range fsMap {
+		// UnFreeze the filesystems, subvolumes and subvolume groups
+		data, err := fm.ReleaseFSQuiesce(ctx, vg.RequestName)
+		if err != nil {
+			log.ErrorLog(ctx, "failed to release filesystem quiesce: %v", err)
+			uErr := cs.deleteSnapshotsAndUndoReservation(ctx, vgs, cr, fsMap, req.GetSecrets())
+			if uErr != nil {
+				log.ErrorLog(ctx, "failed to delete snapshot and undo reservation: %v", uErr)
+			}
+
+			return nil, status.Errorf(codes.Internal, "failed to release filesystem quiesce: %v", err)
+		}
+		state := core.GetQuiesceState(data.State)
+		if state != core.Released {
+			return nil, status.Errorf(codes.Internal, "quiesce operation is in %s state", state)
+		}
+	}
+	resp, rErr := cs.createSnapshotAddToVolumeGroupJournal(ctx,
+		req,
+		true,
+		vg,
+		vgs,
+		cr,
+		fsMap,
+		req.GetSecrets())
+	if rErr != nil {
+		log.ErrorLog(ctx, "failed to create snapshot and add to volume group journal: %v", rErr)
+		if !errors.Is(rErr, cerrors.ErrQuiesceInProgress) {
+			uErr := cs.deleteSnapshotsAndUndoReservation(ctx, vgs, cr, fsMap, req.GetSecrets())
+			if uErr != nil {
+				log.ErrorLog(ctx, "failed to delete snapshot and undo reservation: %v", uErr)
+			}
+		}
+
+		return nil, status.Errorf(codes.Internal,
+			"failed to create snapshot and add to volume group journal: %v",
+			rErr)
+	}
+
+	response := &csi.CreateVolumeGroupSnapshotResponse{}
+	response.GroupSnapshot = &csi.VolumeGroupSnapshot{
+		GroupSnapshotId: vgs.VolumeGroupSnapshotID,
+	}
+	response.GroupSnapshot.ReadyToUse = true
+	response.GroupSnapshot.CreationTime = timestamppb.New(time.Now())
+
+	for _, r := range *resp {
+		r.Snapshot.GroupSnapshotId = vgs.VolumeGroupSnapshotID
+		response.GroupSnapshot.Snapshots = append(response.GroupSnapshot.Snapshots, r.Snapshot)
+	}
+
+	return response, nil
+}
+
+// createSnapshotAddToVolumeGroupJournal creates the snapshot and adds the
+// snapshotID and volumeID to the volume group journal omap. If the freeze is
+// true then it will freeze the subvolumes and subvolume groups before creating
+// the snapshot and unfreeze them after creating the snapshot. If the freeze is
+// false it will call createSnapshot and get the snapshot details for the
+// volume and add the snapshotID and volumeID to the volume group journal omap.
+// If any error occurs other than ErrInProgress it will delete the snapshots
+// and undo the reservation and return the error.
+func (cs *ControllerServer) createSnapshotAddToVolumeGroupJournal(
+	ctx context.Context,
+	req *csi.CreateVolumeGroupSnapshotRequest,
+	freeze bool,
+	vgo *store.VolumeGroupOptions,
+	vgs *store.VolumeGroupSnapshotIdentifier,
+	cr *util.Credentials,
+	fsMap map[string]core.FSQuiesceClient,
+	secrets map[string]string) (
```

Comment:
Wow, one of the previous functions has 6 parameters, and I considered mentioning that. This one has :eight:! It becomes rather unmanageable now... Maybe it makes sense to split the functionality of this function into smaller pieces to enhance the maintainability.

Reply:
Removed one parameter for now, Taking this as action items to refactor it, it looks like its not easy to refactor as the internal function needs more parameter

Reply:
Refactored now to remove 2 parameters, now both will have 6 but will try to reduce it (if possible)